lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+        /// order shutdown.
+        /// </summary>
+        /// <param name="value"></param>
         public void OnNext(IFailedEvaluator value)
         {
-            Logger.Log(Level.Info, "An evaluator failed, checking if it failed before context and service was submitted");
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-
-            if (value.FailedContexts != null && value.FailedContexts.Count != 0)
+            if (_areImruTasksCompleted)
             {
-                Logger.Log(Level.Info, "Some active context failed, cannot continue IMRU task");        
-                Exceptions.Throw(new Exception(), Logger);
+                Logger.Log(Level.Info,
+                    string.Format("Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
             }
 
+            Logger.Log(Level.Info,
+                string.Format("Evaluator with Id: {0} failed with Exception: {1}", value.Id, value.EvaluatorException));
+            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
             if (currFailedEvaluators > _allowedFailedEvaluators)
             {
                 Exceptions.Throw(new Exception("Cannot continue IMRU job, Failed evaluators reach maximum limit"),
                     Logger);
             }
 
-            Logger.Log(Level.Info, "Requesting for the failed evaluator again");
-
-            _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
+            _serviceAndContextConfigurationProvider.RecordEvaluatorFailureById(value.Id);
+            bool isMaster = _serviceAndContextConfigurationProvider.IsMasterEvaluatorId(value.Id);
 
-            // if active context stage is reached for Update Task then assume that failed
-            // evaluator belongs to mapper
-            if (_reachedUpdateTaskActiveContext)
+            // If failed evaluator is master then ask for master 
+            // evaluator else ask for mapper evaluator
+            if (!isMaster)
             {
+                Logger.Log(Level.Info, "Requesting for the failed map evaluator again");
                 RequestMapEvaluators(1);
             }
             else
             {
+                Logger.Log(Level.Info, "Requesting for the failed master evaluator again");
                 RequestUpdateEvaluator();
             }
         }
 
         /// <summary>
-        /// Specfies how to handle exception or error
+        /// Specified what to do if Failed Context is received.
+        /// An exception is thrown if tasks are not completed.
         /// </summary>
-        /// <param name="error">Kind of exception</param>
-        public void OnError(Exception error)
+        /// <param name="value"></param>
+        public void OnNext(IFailedContext value)
+        {
+            if (_areImruTasksCompleted)
+            {
+                Logger.Log(Level.Info,
```

Comment:
As discussed these bits of IMRU are going to be replaced very soon. No point writing tests